compositor_pipeline/src/queue/video_queue.rs

```diff
@@ -98,68 +120,175 @@ impl VideoQueue {
+                    // regardless if they are there or not.
+                    offset > next_buffer_pts
+                }
+                None => {
+                    // It represent stream that still buffering. We now that frames
+                    // from this input will not be used for this batch, so it is fine
+                    // to consider this "ready".
+                    true
+                }
+            };
+        };
+
+        fn has_frame_for_pts(queue: &VecDeque<Frame>, next_buffer_pts: Duration) -> bool {
             match queue.back() {
                 Some(last_frame) => last_frame.pts >= next_buffer_pts,
                 None => false,
             }
         }
 
-        while !is_ready(&self.queue, next_buffer_pts) {
-            if self.try_enqueue_frame(clock_start).is_err() {
+        while !has_frame_for_pts(&self.queue, next_buffer_pts) {
+            if self.try_enqueue_frame().is_err() {
                 return false;
             }
         }
         true
     }
 
-    fn try_enqueue_frame(&mut self, clock_start: Instant) -> Result<(), TryRecvError> {
+    /// Drops frames that won't be used.if oldest pts that we will need in the future is
+    /// `next_buffer_pts`.
+    ///
+    /// Finds frame that is closest to the next_buffer_pts and removes everything older.
+    /// Frames in queue have monotonically increasing pts, so we can just drop all the frames
+    /// before the "closest" one.
+    /// If dropping frames removes everything from the queue try to enqueue some new frames
+    /// and repeat the process.
+    fn drop_old_frames(&mut self, next_buffer_pts: Duration, queue_start: Instant) {
+        let Some(next_buffer_pts) = self.input_pts_from_queue_pts(next_buffer_pts, queue_start)
+        else {
+            // before first frame so nothing to drop
+            return;
+        };
+
+        let next_output_buffer_nanos = next_buffer_pts.as_nanos();
+
+        loop {
+            let closest_diff_frame_index = self
+                .queue
+                .iter()
+                .enumerate()
+                .min_by_key(|(_index, frame)| {
+                    frame.pts.as_nanos().abs_diff(next_output_buffer_nanos)
+                })
+                .map(|(index, _frame)| index);
+
+            if let Some(index) = closest_diff_frame_index {
+                self.queue.drain(0..index);
+            }
+
+            if !self.queue.is_empty() {
+                return;
+            }
+
+            // if queue is empty then try to enqueue some more frames
+            if self.try_enqueue_frame().is_err() {
+                return;
+            }
+        }
+    }
+
+    /// Calculate input pts based on queue pts and queue start time. It can trigger
+    /// enqueue internally.
+    ///
+    /// Returns None if:
+    /// - Input is not ready and offset is unknown
+    /// - If offset is negative (PTS refers to moment from before stream start)
+    fn input_pts_from_queue_pts(
```

Comment:
I don't like the idea of triggering `enqueue` internally in such functions. Shouldn't this be done separately? The current implementation works with the assumption, that this function does that, while the name `input_pts_from_queue_pts` doesn't suggest that at all

Reply:
It depends on how you think about it. Yes if described this way this seems like anit-pattern. I would prefer cleaner approach, but I don't think this can be done better without sacrificing some behaviors or type-safety.

I recommend thinking about it this way:
- data is enqueued when it's sent via channel
- queue is represented by 2 items self.queue for already evaluated data and self.receiver for lazy data

Reply:
I agree that when thinking this way it makes sense. I think we can better integrate this information into a type system.

If we think about it this way, can't we do sth like:
```rust
struct PayloadQueue<T: Payload> {
    enqueued: VecDeque<T>,
    receiver: Receiver<T>
}

impl<T> PayloadQueue<T> {
   ...
   pub fn get_ref(&self) -> &VecDeque<T> {
        try_enqueue(...);
        &self.enqueued
   }
}
```
and guarantee on a type system that whenever some part of the code accesses enqueued data it first checks the receiver.

Reply:
- This would not be correct, we don't want to enqueue every time we get a ref. 
- To create that abstraction we would also need to have InputProcessor there.
- It's possible to create PayloadQueue abstraction that would work, but this would move most of the logic inside PayloadQueue. As a result PayloadQueue  would be type safe, but all the complicated logic would be inside making this type safety unhelpful